storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/ConstraintSolverStrategy.java

```diff
@@ -613,14 +743,51 @@ public void backtrack(Map<ExecutorDetails, String> execToComp, RasNode node, Wor
             String comp = execToComp.get(exec);
             LOG.trace("Backtracking {} {} from {}", exec, comp, workerSlot);
             if (okToRemoveFromWorker[execIndex]) {
-                workerCompAssignment.get(workerSlot).remove(comp);
+                Map<String, Integer> oneMap = workerCompAssignmentCnts.get(workerSlot);
+                oneMap.put(comp, oneMap.getOrDefault(comp, 0) - 1); // decrement assignment count
                 okToRemoveFromWorker[execIndex] = false;
             }
             if (okToRemoveFromNode[execIndex]) {
-                nodeCompAssignment.get(node).remove(comp);
+                Map<String, Integer> oneMap = nodeCompAssignmentCnts.get(node);
+                oneMap.put(comp, oneMap.getOrDefault(comp, 0) - 1); // decrement assignment count
                 okToRemoveFromNode[execIndex] = false;
             }
             node.freeSingleExecutor(exec, td);
         }
+
+        /**
+         * Use this method to log the current component assignments on the Node.
+         * Useful for debugging and tests.
+         */
+        public void logNodeCompAssignments() {
+            if (nodeCompAssignmentCnts == null || nodeCompAssignmentCnts.isEmpty()) {
+                LOG.info("NodeCompAssignment is empty");
+                return;
+            }
+            StringBuffer sb = new StringBuffer();
+            int cntAllNodes = 0;
+            int cntFilledNodes = 0;
+            for (RasNode node: new TreeSet<RasNode>(nodeCompAssignmentCnts.keySet())) {
```

Comment:
`new TreeSet<RasNode>` can be `new TreeSet<>`

Reply:
fixed